src/lib.rs

```diff
@@ -423,7 +423,9 @@ impl<T: Debug + PrimInt + One + Zero> Vob<T> {
         Range{start, end}
     }
 
-    /// Returns an iterator which produces the index of each set bit in the specified range.
+    /// Returns an iterator which efficiently produces the index of each set bit in the specified
+    /// range. Assuming appropriate support from your CPU, this is much more efficient than
```

Comment:
Did you disassemble a binary to check that rustc uses vector instructions?

Reply:
I'm thinking of things like x86's `popcnt`: not every CPU, I think, has such an instruction. It makes things *much* faster if you have even mildly sparse Vobs. I don't know whether things get vectorised in this case: I doubt it very much, to be honest, as I suspect there's too much complex code going on for the compiler to manage to vectorise things.

Reply:
So, I wonder why we are making these claims.

If you do a quick `objdump -D` and look at the code, you can dispel any doubt.

Reply:
Because x86 has `popcnt` and it makes some of my code 10x faster than doing `iter.filter()`.

Reply:
Think of it another way: if I didn't tell you that `iter_set_bits` was efficient in many cases, why would you bother using it?

Reply:
There seems to be confusion. I'm asking: are you certain rustc emitted `popcnt` into your binaries?

If so, great, this is OK. If not, then the docs are wrong.

Reply:
The doc comment makes no claims about `popcnt` -- I'm doing that to give you an intuition about why this stuff works :) Ultimately I depend entirely on what things like `trailing_zeros` in Rust's standard library use, which is why I added the "depending on processor support" qualification.

What I can tell you definitively is that I've measured `iter_set_bits` as being *much* faster than doing the semantically equivalent `iter().enumerate().filter(|(b, _)| b)`.

Reply:
Saying it's faster is fine.

If we are certain that rustc uses the CPU extensions, then OK.

[FWIW, there's no use of popcnt in your rlib on release/openbsd]

Reply:
OK, I admit to taking a shortcut: `popcnt` is a shortcut for a family of instructions like `tzcnt` (which, I guess, is used somewhere!).

Reply:
Afraid not:
```
arrakis:vob> for i in /usr/local/lib/rustlib/x86_64-unknown-openbsd/lib/*.so; do objdump -D $i | grep tzcnt; done
arrakis:vob> 
```

After some inspection, `trailing_zeroes` is implemented as a call to an LLVM intrinsic called `cttz`:
 * https://doc.rust-lang.org/src/core/num/mod.rs.html#1356-1358
 * https://github.com/rust-lang/rust/blob/master/src/librustc_trans/intrinsic.rs
 * https://llvm.org/docs/LangRef.html#llvm-cttz-intrinsic

So Rust has no explicit optimisations per-se. It's at the mercy of LLVM to do a good job. I couldn't find the implementation of `cttz`.

Unwinding, I'm less interested in talking about specific instructions, but I wonder if you should just say it's efficient and leave it at that?

Reply:
Ah here's the implementation for `cttz` on x86_64 in LLVM 6 (IntrinsicLowering.cpp):
```
  case Intrinsic::cttz: {
    // cttz(x) -> ctpop(~X & (X-1))
    Value *Src = CI->getArgOperand(0);
    Value *NotSrc = Builder.CreateNot(Src);
    NotSrc->setName(Src->getName() + ".not");
    Value *SrcM1 = ConstantInt::get(Src->getType(), 1);
    SrcM1 = Builder.CreateSub(Src, SrcM1);
    Src = LowerCTPOP(Context, Builder.CreateAnd(NotSrc, SrcM1), CI);
    CI->replaceAllUsesWith(Src);
    break;
  }
```

But curiously there are no `ctpop` instructions in any of the libs shipped with Rust.

Reply:
Yes, it was always going to be at LLVM's mercy. But the point is that LLVM is then at your processor's mercy. If you don't have fast processor support for things like trailing_zeros, `iter_set_bits` won't be fast. So I can't make a guarantee about *all* platforms which is why I said "processor support".

Reply:
I think you should remove the part about CPU features. Bencher8's CPU supports all of the features we are talking about, yet Rust isn't using them.

[If you disagree, then I'll fold, because this is taking too long at this point.]

Reply:
We've definitely gone too deep into this ;) but it turns out that on x86 LLVM is using the `bsf` instruction. I didn't know about that; it's a more generalised version of `popcnt` and the like. LLVM may also be using `bsf` because some recent x86 chips have a bug in `popcnt` (an incorrect dependency between the registers used) and friends that makes them slower.

Anyway, I think we've probably done this to death ;)

Reply:
Now we are getting somewhere! All I've been asking is proof that your docs are accurate!

```
vext01@bencher8:~$ for i in ~/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/*.so; do objdump -D $i | grep bsf; done
    47dc:	0f bc 76 a0          	bsf    -0x60(%rsi),%esi
    1220:	0f bc 67 a3          	bsf    -0x5d(%rdi),%esp
    4a19:	0f bc b1 af d4 e7 de 	bsf    -0x21182b51(%rcx),%esi
     d5e:	49 0f bc b7 c7 d7 d7 	bsf    0x3cd7d7c7(%r15),%rsi
    982d:	0f bc 8a f8 ec 86 d2 	bsf    -0x2d791308(%rdx),%ecx
  6bf73e:	0f bc ce             	bsf    %esi,%ecx
...
```

Given this, OK :)